## TASK 1

In [15]:
import pandas as pd
import numpy as np

# Read dataset
df = pd.read_csv("cosmetics.csv", engine="python", on_bad_lines="skip")# only one bad line that is 1338th row
# Display 5 random rows
df.sample(5)

,Label,Brand,Name,Price,Rank,Ingredients,Combination,Dry,Normal,Oily,Sensitive
1037,Face Mask,ORIGINS,RitualiTea™ Matcha Madness™ Revitalizing Powde...,36,3.8,"Avena Sativa (Oat) Kernel Meal, Sodium Bicarbo...",1,1,1,1,0
1253,Eye cream,DIOR,Capture Totale 360 Light-Up Open-Up Replenishi...,110,4.1,Visit the Dior boutique,0,0,0,0,0
898,Face Mask,DR. JART+,Shake & Shot™ Rubber Masks,12,4.1,-Vitamin C (Ascorbic Acid): Brightens the look...,1,1,1,1,1
1067,Face Mask,ORIGINS,Flower Fusion™ Lavender Soothing Sheet Mask,7,4.3,"Water, Butylene Glycol, Bis-Peg-18 Methyl Ethe...",1,1,1,1,1
437,Cleanser,CLINIQUE,Rinse-Off Eye Makeup Solvent,19,4.3,"Water , Peg-32 , Butylene Glycol , Disodium Co...",0,0,0,0,0


In [12]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1337 entries, 0 to 1336
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Label        1337 non-null   str    
 1   Brand        1337 non-null   str    
 2   Name         1337 non-null   str    
 3   Price        1337 non-null   int64  
 4   Rank         1337 non-null   float64
 5   Ingredients  1337 non-null   str    
 6   Combination  1337 non-null   int64  
 7   Dry          1337 non-null   int64  
 8   Normal       1337 non-null   int64  
 9   Oily         1337 non-null   int64  
 10  Sensitive    1337 non-null   int64  
dtypes: float64(1), int64(6), str(4)
memory usage: 1.1 MB


In [13]:
df.columns

Index(['Label', 'Brand', 'Name', 'Price', 'Rank', 'Ingredients', 'Combination',
       'Dry', 'Normal', 'Oily', 'Sensitive'],
      dtype='str')

In [14]:
df["Label"].value_counts()

Label
Moisturizer    298
Cleanser       281
Face Mask      266
Treatment      248
Eye cream      209
Sun protect     35
Name: count, dtype: int64

In [16]:
!pip freeze > requirements.txt

## TASK 2

In [19]:
# Keep only moisturizers
moisturizers = df[df["Label"] == "Moisturizer"]

# View first 5 rows
moisturizers.head()

,Label,Brand,Name,Price,Rank,Ingredients,Combination,Dry,Normal,Oily,Sensitive
0,Moisturizer,LA MER,Crème de la Mer,175,4.1,"Algae (Seaweed) Extract, Mineral Oil, Petrolat...",1,1,1,1,1
1,Moisturizer,SK-II,Facial Treatment Essence,179,4.1,"Galactomyces Ferment Filtrate (Pitera), Butyle...",1,1,1,1,1
2,Moisturizer,DRUNK ELEPHANT,Protini™ Polypeptide Cream,68,4.4,"Water, Dicaprylyl Carbonate, Glycerin, Ceteary...",1,1,1,1,0
3,Moisturizer,LA MER,The Moisturizing Soft Cream,175,3.8,"Algae (Seaweed) Extract, Cyclopentasiloxane, P...",1,1,1,1,1
4,Moisturizer,IT COSMETICS,Your Skin But Better™ CC+™ Cream with SPF 50+,38,4.1,"Water, Snail Secretion Filtrate, Phenyl Trimet...",1,1,1,1,1


In [23]:
# Keep moisturizers suitable for dry skin
moisturizers_dry = moisturizers[moisturizers["Dry"] == 1]

# Reset index
moisturizers_dry = moisturizers_dry.reset_index(drop=True)

# Check size
moisturizers_dry

,Label,Brand,Name,Price,Rank,Ingredients,Combination,Dry,Normal,Oily,Sensitive
0,Moisturizer,LA MER,Crème de la Mer,175,4.1,"Algae (Seaweed) Extract, Mineral Oil, Petrolat...",1,1,1,1,1
1,Moisturizer,SK-II,Facial Treatment Essence,179,4.1,"Galactomyces Ferment Filtrate (Pitera), Butyle...",1,1,1,1,1
2,Moisturizer,DRUNK ELEPHANT,Protini™ Polypeptide Cream,68,4.4,"Water, Dicaprylyl Carbonate, Glycerin, Ceteary...",1,1,1,1,0
3,Moisturizer,LA MER,The Moisturizing Soft Cream,175,3.8,"Algae (Seaweed) Extract, Cyclopentasiloxane, P...",1,1,1,1,1
4,Moisturizer,IT COSMETICS,Your Skin But Better™ CC+™ Cream with SPF 50+,38,4.1,"Water, Snail Secretion Filtrate, Phenyl Trimet...",1,1,1,1,1
...,...,...,...,...,...,...,...,...,...,...,...
185,Moisturizer,KIEHL'S SINCE 1851,Ultra Facial Deep Moisture Balm,29,4.7,"Water, Glycerin, Shea Butter, Glyceryl Stearat...",0,1,1,0,0
186,Moisturizer,SHISEIDO,White Lucent All Day Brightener Broad Spectrum...,62,4.6,"Water, Sd Alcohol 40-B, Dimethicone, Dipropyle...",1,1,1,0,0
187,Moisturizer,SATURDAY SKIN,Featherweight Daily Moisturizing Cream,49,4.6,"Water, Butylene Glycol, Ethylhexyl Palmitate, ...",1,1,1,1,1
188,Moisturizer,KATE SOMERVILLE,Goat Milk Moisturizing Cream,65,4.1,"Water, Ethylhexyl Palmitate, Myristyl Myristat...",1,1,1,1,1


In [26]:
moisturizers_dry[["Brand","Name","Price","Rank"]].head(10)

,Brand,Name,Price,Rank
0,LA MER,Crème de la Mer,175,4.1
1,SK-II,Facial Treatment Essence,179,4.1
2,DRUNK ELEPHANT,Protini™ Polypeptide Cream,68,4.4
3,LA MER,The Moisturizing Soft Cream,175,3.8
4,IT COSMETICS,Your Skin But Better™ CC+™ Cream with SPF 50+,38,4.1
5,DRUNK ELEPHANT,Lala Retro™ Whipped Cream,60,4.2
6,DRUNK ELEPHANT,Virgin Marula Luxury Facial Oil,72,4.4
7,KIEHL'S SINCE 1851,Ultra Facial Cream,29,4.4
8,KIEHL'S SINCE 1851,Midnight Recovery Concentrate,47,4.4
9,SUNDAY RILEY,Luna Sleeping Night Oil,105,4.1


## TASK 3

In [28]:
# Create a working copy
moisturizers = moisturizers_dry.copy()

In [29]:
# Convert ingredient text to lowercase as Machine learning treats different cases as different words, so we standardize everything.
moisturizers["Ingredients"] = moisturizers["Ingredients"].str.lower()

In [30]:
# Split ingredients into a list
moisturizers["Tokens"] = moisturizers["Ingredients"].str.split(", ")

In [43]:
moisturizers["Tokens"].iloc[0]

['algae (seaweed) extract',
 'mineral oil',
 'petrolatum',
 'glycerin',
 'isohexadecane',
 'microcrystalline wax',
 'lanolin alcohol',
 'citrus aurantifolia (lime) extract',
 'sesamum indicum (sesame) seed oil',
 'eucalyptus globulus (eucalyptus) leaf oil',
 'sesamum indicum (sesame) seed powder',
 'medicago sativa (alfalfa) seed powder',
 'helianthus annuus (sunflower) seedcake',
 'prunus amygdalus dulcis (sweet almond) seed meal',
 'sodium gluconate',
 'copper gluconate',
 'calcium gluconate',
 'magnesium gluconate',
 'zinc gluconate',
 'magnesium sulfate',
 'paraffin',
 'tocopheryl succinate',
 'niacin',
 'water',
 'beta-carotene',
 'decyl oleate',
 'aluminum distearate',
 'octyldodecanol',
 'citric acid',
 'cyanocobalamin',
 'magnesium stearate',
 'panthenol',
 'limonene',
 'geraniol',
 'linalool',
 'hydroxycitronellal',
 'citronellol',
 'benzyl salicylate',
 'citral',
 'sodium benzoate',
 'alcohol denat.',
 'fragrance.']

In [36]:
# #Python creates:

# Ingredient
# Index
# water 0
# glycerin 1
# aloe 2
# vitamin e 3

# This dictionary is called ingredient_idx.
# It maps every ingredient to a unique number.
    
ingredient_idx = {}
idx = 0

for ingredients in moisturizers["Tokens"]:
    for ingredient in ingredients:
        if ingredient not in ingredient_idx:
            ingredient_idx[ingredient] = idx
            idx += 1

In [37]:
print("Total unique ingredients:", len(ingredient_idx))

Total unique ingredients: 2233


In [45]:
list(ingredient_idx.items())[:30]

[('algae (seaweed) extract', 0),
 ('mineral oil', 1),
 ('petrolatum', 2),
 ('glycerin', 3),
 ('isohexadecane', 4),
 ('microcrystalline wax', 5),
 ('lanolin alcohol', 6),
 ('citrus aurantifolia (lime) extract', 7),
 ('sesamum indicum (sesame) seed oil', 8),
 ('eucalyptus globulus (eucalyptus) leaf oil', 9),
 ('sesamum indicum (sesame) seed powder', 10),
 ('medicago sativa (alfalfa) seed powder', 11),
 ('helianthus annuus (sunflower) seedcake', 12),
 ('prunus amygdalus dulcis (sweet almond) seed meal', 13),
 ('sodium gluconate', 14),
 ('copper gluconate', 15),
 ('calcium gluconate', 16),
 ('magnesium gluconate', 17),
 ('zinc gluconate', 18),
 ('magnesium sulfate', 19),
 ('paraffin', 20),
 ('tocopheryl succinate', 21),
 ('niacin', 22),
 ('water', 23),
 ('beta-carotene', 24),
 ('decyl oleate', 25),
 ('aluminum distearate', 26),
 ('octyldodecanol', 27),
 ('citric acid', 28),
 ('cyanocobalamin', 29)]

## TASK 4 

In [46]:
# Number of products
M = len(moisturizers)

# Number of unique ingredients
N = len(ingredient_idx)

print("Products:", M)
print("Ingredients:", N)

Products: 190
Ingredients: 2233


In [48]:
# Create empty Document-Term Matrix
A = np.zeros((M, N))

In [49]:
A.shape

(190, 2233)

In [50]:
A

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], shape=(190, 2233))

## TASK 5

In [51]:
ingredient_idx

{'algae (seaweed) extract': 0,
 'mineral oil': 1,
 'petrolatum': 2,
 'glycerin': 3,
 'isohexadecane': 4,
 'microcrystalline wax': 5,
 'lanolin alcohol': 6,
 'citrus aurantifolia (lime) extract': 7,
 'sesamum indicum (sesame) seed oil': 8,
 'eucalyptus globulus (eucalyptus) leaf oil': 9,
 'sesamum indicum (sesame) seed powder': 10,
 'medicago sativa (alfalfa) seed powder': 11,
 'helianthus annuus (sunflower) seedcake': 12,
 'prunus amygdalus dulcis (sweet almond) seed meal': 13,
 'sodium gluconate': 14,
 'copper gluconate': 15,
 'calcium gluconate': 16,
 'magnesium gluconate': 17,
 'zinc gluconate': 18,
 'magnesium sulfate': 19,
 'paraffin': 20,
 'tocopheryl succinate': 21,
 'niacin': 22,
 'water': 23,
 'beta-carotene': 24,
 'decyl oleate': 25,
 'aluminum distearate': 26,
 'octyldodecanol': 27,
 'citric acid': 28,
 'cyanocobalamin': 29,
 'magnesium stearate': 30,
 'panthenol': 31,
 'limonene': 32,
 'geraniol': 33,
 'linalool': 34,
 'hydroxycitronellal': 35,
 'citronellol': 36,
 'benzyl 

In [52]:
def oh_encoder(tokens):

    # Create empty vector
    x = np.zeros(N)

    # Fill 1 where ingredient exists
    for ingredient in tokens:
        if ingredient in ingredient_idx:
            x[ingredient_idx[ingredient]] = 1

    return x

In [71]:
sample_tokens = moisturizers["Tokens"].iloc[78]

encoded = oh_encoder(sample_tokens)

print(encoded[:1000])

[0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.
 0. 0. 0. 0. 0. 0. 0. 0. 1. 1. 1. 0. 1. 0. 1. 0. 1. 0. 0. 1. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0.
 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 1. 0. 0. 0. 0. 0. 1. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.

In [72]:
print("Ingredients in this product:", int(encoded.sum()))

Ingredients in this product: 60


## TASK 6

In [73]:
# Fill the Document-Term Matrix
# enumerate() gives you both the row number (i) and the ingredient list (tokens).
# oh_encoder(tokens) converts ingredients into 1s and 0s.
# A[i] stores that binary vector in the matrix.

for i, tokens in enumerate(moisturizers["Tokens"]):
    A[i] = oh_encoder(tokens)

print("Matrix generated successfully!")

Matrix generated successfully!


In [74]:
A.shape

(190, 2233)

In [79]:
#[product][ingredients list]
A[45][:300]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 1., 0., 0., 0., 0., 1.,
       0., 1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 1., 0.,
       0., 0., 1., 0., 1., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0.

## TASK 7

What is t-SNE?

Right now every moisturizer has 2233 features.
Humans can't visualize 2233 dimensions.
t-SNE compresses them into just 2 coordinates while keeping similar products close together.

In [82]:
#Import TSNE
from sklearn.manifold import TSNE

In [83]:
tsne = TSNE(
    n_components=2,
    learning_rate=200,
    random_state=42,
    init="random"
)

In [84]:
tsne_result = tsne.fit_transform(A)

In [96]:
moisturizers["X"] = tsne_result[:, 0]
moisturizers["Y"] = tsne_result[:, 1]

moisturizers[["Brand","Name","Price","Rank", "X","Y"]].head()

,Brand,Name,Price,Rank,X,Y
0,LA MER,Crème de la Mer,175,4.1,-0.391624,-6.365034
1,SK-II,Facial Treatment Essence,179,4.1,-3.159598,-4.858387
2,DRUNK ELEPHANT,Protini™ Polypeptide Cream,68,4.4,-3.690483,-4.325043
3,LA MER,The Moisturizing Soft Cream,175,3.8,-0.905349,-7.083622
4,IT COSMETICS,Your Skin But Better™ CC+™ Cream with SPF 50+,38,4.1,-6.146412,-1.742621


## TASK 8 
Interactive Visualization using Bokeh

-Plot every moisturizer as a dot.

-Each dot = one product

-Nearby dots = similar ingredients

In [97]:
from bokeh.plotting import figure, show
from bokeh.models import ColumnDataSource, HoverTool

In [98]:
source = ColumnDataSource(moisturizers)

In [99]:
plot = figure(
    width=800,
    height=600,
    title="Moisturizers for Dry Skin (t-SNE Visualization)",
    tools="pan,wheel_zoom,reset,save"
)
# #What are these tools?
# Tool            Purpose
# Pan             Drag graph
# Wheel Zoom      Zoom in/out
# Reset           Return to original
# Save            Download image

In [100]:
plot.scatter(
    x="X",
    y="Y",
    source=source,
    size=8,
    alpha=0.7
)

GlyphRenderer(id='p1133', ...)

## TASK 9

In [114]:
hover = HoverTool(tooltips=[
    ("Index", "$index"),
    ("Brand", "@Brand"),
    ("Product", "@Name"),
    ("Price", "@Price"),
    ("Rating", "@Rank")
])

plot.add_tools(hover)

## TASK 10

In [115]:
show(plot)

## TASK 11

In [116]:
moisturizers[["Brand","Name"]].head(10)

,Brand,Name
0,LA MER,Crème de la Mer
1,SK-II,Facial Treatment Essence
2,DRUNK ELEPHANT,Protini™ Polypeptide Cream
3,LA MER,The Moisturizing Soft Cream
4,IT COSMETICS,Your Skin But Better™ CC+™ Cream with SPF 50+
5,DRUNK ELEPHANT,Lala Retro™ Whipped Cream
6,DRUNK ELEPHANT,Virgin Marula Luxury Facial Oil
7,KIEHL'S SINCE 1851,Ultra Facial Cream
8,KIEHL'S SINCE 1851,Midnight Recovery Concentrate
9,SUNDAY RILEY,Luna Sleeping Night Oil


In [127]:
import re

# Clean and tokenize ingredients
moisturizers["Tokens"] = moisturizers["Ingredients"].apply(
    lambda text: [
        re.sub(r'[\*\.]+$', '', ing.strip().lower())   # remove trailing *, ** and .
        for ing in text.split(",")
        if ing.strip()
    ]
)

In [128]:
product1 = moisturizers.iloc[32]
product2 = moisturizers.iloc[80]
product3 = moisturizers.iloc[75]

print("PRODUCT 1")
print(product1["Brand"])
print(product1["Name"])

print("\nPRODUCT 2")
print(product2["Brand"])
print(product2["Name"])

print("\nPRODUCT 3")
print(product2["Brand"])
print(product2["Name"])

PRODUCT 1
SUNDAY RILEY
U.F.O. Ultra-Clarifying Face Oil

PRODUCT 2
OLEHENRIKSEN
Invigorating Night Transformation™ Gel

PRODUCT 3
OLEHENRIKSEN
Invigorating Night Transformation™ Gel


In [129]:
print("Ingredients of Product 1:\n")
print(product1["Ingredients"])

print("\n" + "="*80 + "\n")

print("Ingredients of Product 2:\n")
print(product2["Ingredients"])

print("\n" + "="*80 + "\n")

print("Ingredients of Product 3:\n")
print(product3["Ingredients"])

Ingredients of Product 1:

silybum marianum (milk thistle) seed oil, nigella sativa (black cumin) seed oil, cucumis sativus (cucumber) seed oil, diisopropyl sebacate, punica granatum (pomegranate) seed oil, vaccinium macrocarpon (cranberry) seed oil, dimethyl isosorbide, linum usitatissimum (flax) seed oil, caprylic/capric triglycerides, ethyl linoleate, hexylresorcinol, glycyrrhiza glabra (licorice) root extract, salicylic acid, prunus armeniaca (apricot) kernel oil, chamomilla recutita flower oil, citrus aurantium amara (neroli) flower oil, melaleuca alternifolia (tea tree) leaf oil, helianthus annuus seed oil, rosmarinus officinalis (rosemary) leaf extract, citrus paradisii (grapefruit) peel oil, melia azadirachta (neem) leaf extract, melia azadirachta (neem) flower extract, coccinia indica fruit extract, amino esters-1, amber leaf extract, solanum melongena (eggplant) fruit extract, aloe barbadensis flower extract, lawsonia inermis (henna) extract, ocimum sanctum leaf extract, ocim

In [130]:
# ---------- Similarity Comparison (Products 32, 75, 80) ----------

pairs = [(32, 75), (32, 80), (75, 80)]

for idx1, idx2 in pairs:
    set1 = set(moisturizers.loc[idx1, "Tokens"])
    set2 = set(moisturizers.loc[idx2, "Tokens"])

    common = sorted(set1.intersection(set2))

    print("="*90)
    print(f"Product {idx1}: {moisturizers.loc[idx1, 'Brand']} - {moisturizers.loc[idx1, 'Name']}")
    print(f"Product {idx2}: {moisturizers.loc[idx2, 'Brand']} - {moisturizers.loc[idx2, 'Name']}")
    print(f"\nNumber of Common Ingredients: {len(common)}\n")

    for ingredient in common:
        print("-", ingredient)

    print("\n")

Product 32: SUNDAY RILEY - U.F.O. Ultra-Clarifying Face Oil
Product 75: JOSIE MARAN - Whipped Argan Oil Face Butter

Number of Common Ingredients: 0



Product 32: SUNDAY RILEY - U.F.O. Ultra-Clarifying Face Oil
Product 80: OLEHENRIKSEN - Invigorating Night Transformation™ Gel

Number of Common Ingredients: 1

- glycyrrhiza glabra (licorice) root extract


Product 75: JOSIE MARAN - Whipped Argan Oil Face Butter
Product 80: OLEHENRIKSEN - Invigorating Night Transformation™ Gel

Number of Common Ingredients: 8

- citric acid
- glycerin
- glyceryl caprylate
- glycine soja (soybean) oil
- phenoxyethanol
- potassium sorbate
- tocopherol
- water




## Conclusion 1: Similar Products (Products 32, 75 & 80)

Conclusion: The comparison of Products 32, 75, and 80 showed varying levels of ingredient similarity. Products 75 and 80 shared the highest number of common ingredients, indicating a closely related moisturizing formulation. Product 32 shared comparatively fewer ingredients with the other two, suggesting it belongs to a different botanical oil-based formulation. This supports the t-SNE visualization, where products positioned closer together exhibit greater chemical similarity.

### Note:
Products 32 and 75 showed 0 common ingredients and Products 32 and 80 showed only 1 common ingredient in the token comparison. This occurs because the project uses basic tokenization, where ingredients are matched as exact strings. Minor differences in formatting (extra spaces, */** symbols, singular vs. plural words, or slight naming variations such as caprylic/capric triglycerides vs. caprylic/ capric triglyceride) are treated as different tokens. Therefore, the actual chemical similarity is higher than the reported overlap, and advanced text normalization would be required to identify these equivalent ingredients.

In [119]:
# Compare two dissimilar products

product1 = moisturizers.iloc[128]
product2 = moisturizers.iloc[152]

print("PRODUCT 1")
print(product1["Brand"])
print(product1["Name"])

print("\nPRODUCT 2")
print(product2["Brand"])
print(product2["Name"])

print("\n" + "="*80 + "\n")

print("Ingredients of Product 1:\n")
print(product1["Ingredients"])

print("\n" + "="*80 + "\n")

print("Ingredients of Product 2:\n")
print(product2["Ingredients"])

PRODUCT 1
KORRES
Wild Rose Vitamin C Active Brightening Oil

PRODUCT 2
ESTÉE LAUDER
Advanced Night Repair Intensive Recovery Ampoules


Ingredients of Product 1:

isoamyl laurate, caprylic/capric triglyceride, prunus armeniaca (apricot) kernel oil, ascorbyl tetraisopalmitate, squalane, helianthus annuus (sunflower) seed oil, dimethyl isosorbide, rosa canina fruit oil, rosa gallica flower powder, alcohol denat., water, bisabolol, citronellol, geraniol, glyceryl caprylate, physalis angulata extract, rosa damascena flower oil, rosmarinus officinalis (rosemary) extract, tocopherol, tocopheryl acetate, zingiber officinale (ginger) root extract.


Ingredients of Product 2:

advanced night rpr int rcv ampoule division: el (estee lauder)ingredients: caprylic/capric triglyceride , squalane , aleurites moluccana (kukui) seed oil , prunus armeniaca (apricot) kernel oil , bisabolol , garcinia mangostana peel extract , salicornia herbacea extract , magnolia officinalis bark extract , salvia hispani

In [120]:
# ---------- Dissimilarity Comparison (Products 128, 152) ----------

set1 = set(moisturizers.loc[128, "Tokens"])
set2 = set(moisturizers.loc[152, "Tokens"])

common = sorted(set1.intersection(set2))

print("Number of common ingredients:", len(common))
print("\nCommon Ingredients:")

for ing in common:
    print("-", ing)

Number of common ingredients: 0

Common Ingredients:


## Conclusion 2: Dissimilar Products (Products 128 & 152)

Conclusion: Products 128 and 152 shared 0 common ingredients, indicating completely different chemical formulations. Their lack of ingredient overlap validates the t-SNE mapping, where products located far apart represent highly dissimilar cosmetic compositions. This demonstrates the effectiveness of the content-based recommendation system in distinguishing both similar and dissimilar products.